registry/data-models/models.py

```diff
@@ -0,0 +1,146 @@
+from pydantic import BaseModel
+from typing import List
+
+"""
+This file defines abstract backend data models for feature registry.
+Backend data models will be used by backend API server to talk to feature registry backend.
+Purpose of this is to decouple backend data models from API specific data models.
+For each feature registry provider/implementation, they will extend this abstract
+data models and backend API.
+Diagram of the data models:  ./data-model-diagram.md
+"""
+
+
+class FeatureId(BaseModel):
+    """
+    Id for Feature, it's unique ID represents Feature.
+    Id can be a simple string, int or complex key.
+    """
+    id: str  # id of a feature
+
+
+class FeatureNameId(BaseModel):
+    """
+    Id for FeatureName, it's unique ID represents FeatureName.
+    Id can be a simple string, int or complex key.
+    """
+    id: str  # id of a FeatureName
+
+
+class AnchorId(BaseModel):
+    """
+    Id for Anchor, it's unique ID represents Anchor.
+    Id can be a simple string, int or complex key.
+    """
+    id: str  # id of a anchor
+
+
+class ProjectId(BaseModel):
+    """
+    Id for Project, it's unique ID represents Project.
+    Id can be a simple string, int or complex key.
+    """
+    id: str  # id of a project
+
+
+class Source(BaseModel):
+    """
+    Source of the feature.
+    It defines where the feature is extracted or derived from.
+    """
+    pass
+
+
+class DataSource(Source):
+    """
+    Data source of the feature.
+    It defines the raw data source the feature is extracted from.
+    """
+    pass
+
+
+class FeatureSource(BaseModel):
+    """
+    Represents a feature source for a derived feature. That is, it is a source 'FeatureName' which is used for
+    creating other derived features.
+    """
+    input_feature_name_id: FeatureNameId  # Input feature name Key
+
+
+class MultiFeatureSource(Source):
+    """
+    Feature sources of the feature.
+    It defines one to many features where the feature is derived from.
+    """
+    sources: List[FeatureSource]  # All source features which the feature is derived from
```

Comment:
`MultiFeatureSource` has a very different topology than `FeatureSource`, please figure out how should the lineage be generated/queried.
One typical usage is to check if a feature is an aggregation or non-aggregation feature, the latter cannot be used in a feature generation job. We need a solid abstraction over these features/sources along with required operations/methods to avoid ad-hoc patch everywhere.

Reply:
Gotcha, we can pinpoint it in the next RB.
Perhaps you could provide an example for us?

Reply:
From current code, MultiFeatureSource is just a list of FeatureSource, I added this class just to make it consistent with DataSource. For Lineage, MultiFeatureSource would be one node, and it contains 1 to n FeatureSource with each of them point to a FeatureName.